### ISSUE TITLE
Expected data format and shapes for combining station observations and gridded data

### ISSUE THREAD

**Issue by nilsleh** [cambridge-mlg/gridded-tnp#1]
Hi @cddcam ,

Thanks for the cool work and open-sourcing the code. I was wondering whether you could give an example of a batch of data and their expected dictionary keys and values with shapes etc, to get a better idea of what the `MultiModalGriddedTNPEncoder` expects in its [forward pass](https://github.com/cambridge-mlg/gridded-tnp/blob/e338aaad208b56710008516bb71e1fbf00b8cf66/tnp/models/gridded_tnp.py#L110). Perhaps a python function that creates dummy data accordingly or a print statement, that shows the expected structure. I also had a look [here](https://github.com/cambridge-mlg/gridded-tnp/blob/e338aaad208b56710008516bb71e1fbf00b8cf66/tnp/data/era5/mm_era5.py#L22) for the MM ERA5 Generator. Perhaps, it could be possible for you to upload a single small batch of data so that the code can be expected without the user downloading all the data?

Thanks in advance!

**Comment by nilsleh** [comment]
After further inspection of the code, I came up with the following test file:

```python
import torch
import tempfile
import os
import hiyapyco
from hydra.utils import instantiate
import xarray as xr
import numpy as np
import shutil

from tnp.utils.experiment_utils import deep_convert_dict, extract_config


def create_mock_era5_data(
    data_dir: str,
    data_vars: list,
    time_steps: int = 4,
    lat_size: int = 24,
    lon_size: int = 60,
):
    """Create mock ERA5 data files."""
    base_time = np.datetime64("2018-01-01T00:00:00")
    times = base_time + np.arange(time_steps) * np.timedelta64(1, "h")
    lats = np.linspace(49.0, 25.0, lat_size)
    lons = np.linspace(-125.0, -66.0, lon_size)

    data_arrays = {}
    for var in data_vars:
        data = np.random.randn(time_steps, lat_size, lon_size) * 10 + 20
        data_arrays[var] = xr.DataArray(
            data,
            dims=["valid_time", "latitude", "longitude"],
            coords={"valid_time": times, "latitude": lats, "longitude": lons},
        )

    dataset = xr.Dataset(data_arrays)
    os.makedirs(data_dir, exist_ok=True)
    fname = os.path.join(data_dir, "era5_test_data.nc")
    dataset.to_netcdf(fname, engine="h5netcdf")
    return ["era5_test_data.nc"]


def test_multimodal_tnp_encoder_from_configs():
    """Test MultiModalGriddedTNPEncoder using config files."""

    # Load and merge config files
    config_files = [
        "experiments/configs/models/mm-swintnp.yml",
        "experiments/configs/models/grid_encoders/pt-ge-tt.yml",
        "experiments/configs/models/grid_decoders/mhca-gd.yml",
        "experiments/configs/generators/mm-era5.yml",
    ]

    raw_config = deep_convert_dict(
        hiyapyco.load(
            config_files, method=hiyapyco.METHOD_MERGE, usedefaultyamlloader=True
        )
    )

    # Create mock data
    temp_data_dir = tempfile.mkdtemp()
    data_vars = ["u_1000", "v_1000", "u_850", "v_850", "u_700", "v_700"]
    fnames = create_mock_era5_data(temp_data_dir, data_vars, 4, 24, 60)

    # Configure for testing
    raw_config.setdefault("data", {}).update(
        {
            "data_dir": temp_data_dir,
            "train_fnames": fnames,
            "val_fnames": fnames,
        }
    )

    raw_config.setdefault("misc", {}).update(
        {
            "seed": 42,
            "num_workers": 1,
            "num_val_workers": 1,
            "logging": False,
            "pl_logging": False,
        }
    )

    raw_config.setdefault("params", {}).update(
        {
            "data_vars": data_vars,
            "time_range": ["2018-01-01", "2018-01-01"],
            "val_time_range": ["2018-01-01", "2018-01-01"],
        }
    )

    # Extract and instantiate experiment
    experiment_config, _ = extract_config(raw_config, [])
    experiment = instantiate(experiment_config)

    print(f" Model instantiated: {type(experiment.model)}")
    print(f" Parameters: {sum(p.numel() for p in experiment.model.parameters()):,}")

    # Generate test batch and run forward pass
    test_batch = experiment.generators.train.generate_batch()

    with torch.no_grad():
        if hasattr(test_batch, "time_grid") and test_batch.time_grid is not None:
            output = experiment.model(
                test_batch.xc, test_batch.yc, test_batch.xt, test_batch.time_grid
            )
        else:
            output = experiment.model(test_batch.xc, test_batch.yc, test_batch.xt)

    print(f"✓ Forward pass: {output.mean.shape}")

    # Cleanup
    shutil.rmtree(temp_data_dir, ignore_errors=True)

    return {"model": experiment.model, "output": output, "test_batch": test_batch}


if __name__ == "__main__":
    results = test_multimodal_tnp_encoder_from_configs()
    print(f" Test successful! Output: {results['output'].mean.shape}")
``

**Comment by cddcam** [comment]
Hi @nilsleh,

First of all, thanks for the feedback and for your suggestion for the test file, I will add it into the repo.

Alternatively, you could also use some of the steps from the `README.md` file to download just a couple of months' worth of data, which shouldn't take too long. If you do this, there are indeed certain steps you need to change in the configs. For example:

1. You could run `python experiments/mm_era5_download.py --start_year 2009 --end_year 2009 --num_processes 4` to download just one year of data (or stop it after a few months, or modify the `mm_era5_download.py` script to only download a couple of months)
2. With a few months downloaded, you then need to change the `mm-era5-nc-files.yml` under `experiments/configs/data/mm-era5-nc-files.yml` to reflect the name of the downloaded date range.
3. You also need to specify the correct range in `experiments/configs/generators/mm-era5.yml` under `time_range` (and for `val_time_range`).

You should then be able to run the training script through `lightning_train.py`, as specified in the `README.md`.

Additionally, I can tell you what the data should look like with the setup we have for the multi-modal wind experiment:
- `x_grid` contains all the gridded locations within the area we chose (US - between lat 49 - 25, and long 125 - -66). This accounts for 97 x 237 gridded locations at 0.25°. We also choose T=4 time steps, which is going to be another dimension of `x_grid`. In the end, `x_grid` ends up being `[B, T, lat, lon, 3]`, where `B` is the batch size, and 3 comes from the 3 dimensions: time, latitude, longitude. From `x_grid` we can construct `x` by flattening the inner dimensions to give `[B, T x lat x lon, 3]`.
- `y_grid` contains the values of the `data_vars` specified in `experiments/configs/generators/mm-era5.yml` at the selected gridded locations. Its shape is `[B, T, lat, lon, 6]` in our case because `data_vars` contains 6 variables: `u_1000, u_850, u_700, v_1000, v_850, v_700`. From `y_grid` we can construct `y` by flattening the inner dimensions to give `[B, T x lat x lon, 6]`.
- `xc` represents the context location, and is a dictionary with keys `data_vars`. Each value `xc[key]` is of shape `[B, Nc, 3]`, with `B` the batch size, `Nc` the number of context points equal to `pc*N`, and 3 from the time, latitude, longitude dimensions.
- `yc` represents the context values, and is a dictionary with keys `data_vars`. Each value `yc[key]` is of shape `[B, Nc, 1]`, with `B` the batch size, `Nc` the number of context points equal to `pc*N`, and 3 from the time, latitude, longitude dimensions.
- `xt` represents the target locations and it is a tensor of shape `[B, Nt, 3]`, where `Nt=pt*N.`
- `yt` represents the target values and it is a tensor of shape `[B, Nt, 6]`, where `Nt=pt*N.`

If you have any other questions / feedback, feel free to post it. I will try my best to respond as quickly as possible :)

**Comment by nilsleh** [comment]
@cddcam  Thank you for the elaborate description, this helps a lot! 

Perhaps three more follow up question:
-  the `ERA5MultiModalBatch` has attributes, `x`, `y` and `xc` and `yc`, but `xc` and `yc` are just subsampled versions according to the `pc` parameter, correct? And you just keep `x` and `y` around for plotting utility purposes if I understand correctly, since the model only requires context and target sets.
- with the ERA5 example, the variables `u_1000, u_850, u_700, v_1000, v_850, v_700` are all gridded variables. However, if I understand correctly, station data would follow the same dimensions as `xc`, `yc`, `xt` and `yt`? This would be for having a mix of station and gridded data as context or target sets, similar to the [Deep Sensor](https://alan-turing-institute.github.io/deepsensor/user-guide/task_loader.html#calling-a-taskloader) example.
- how does the Gridded-TNP framework handle gridded data of different resolutions?

**Comment by nilsleh** [comment]
@cddcam Sorry another follow up question, I just noticed that when you said

> * `yc` represents the context values, and is a dictionary with keys `data_vars`. Each value `yc[key]` is of shape `[B, Nc, 1]`, with `B` the batch size, `Nc` the number of context points equal to `pc*N`, and 3 from the time, latitude, longitude dimensions.

The shape of the values `yc[key]` is `[B, Nc, 1]` because that is the dimension of a single of the - in this data example - 6 different data variables, but they are just separated under different keys, and there is no notion of lat, lon, time, correct?

**Comment by cddcam** [comment]
@nilsleh No worries. To address your other questions:
1.  > * the `ERA5MultiModalBatch` has attributes, `x`, `y` and `xc` and `yc`, but `xc` and `yc` are just subsampled versions according to the `pc` parameter, correct? And you just keep `x` and `y` around for plotting utility purposes if I understand correctly, since the model only requires context and target sets.

You are exactly right.

2.  > * with the ERA5 example, the variables `u_1000, u_850, u_700, v_1000, v_850, v_700` are all gridded variables. However, if I understand correctly, station data would follow the same dimensions as `xc`, `yc`, `xt` and `yt`? This would be for having a mix of station and gridded data as context or target sets, similar to the [Deep Sensor](https://alan-turing-institute.github.io/deepsensor/user-guide/task_loader.html#calling-a-taskloader) example.

If you have a mix of station and gridded data, the more relevant example is the station data experiment. There, we predict t2m station measurements by leveraging 1) a subset of t2m station data and 2) skin temperature as the gridded variable. We extracted both variables from the ERA5 reanalysis dataset but t2m is off-the-grid in this case because we define the grid to be a coarsened 180 x 360 grid. So we:
- Take the skin temperature at native resolution and coarsen it to 180 x 360. This is our definition of the "grid".
- Take the t2m temperature from ERA5 at native resolution (721 x 1440). Although this is technically gridded, the finer discretisation means it becomes off-the-grid with respect to the 180 x 360 grid. We then have a set of true station locations and map them to the closest point on the finer 721 x 1440 grid and use the ERA5 measurements as the measurement. 

These are all implemented in `tnp/data/era5/ootg_era5.py` under the relevant generator `ERA5StationOOTGDataGenerator`.

However, you do not need to follow this setup and the locations of the off-the-grid points can be absolutely arbitrary. In our case, this was a convenient choice from a data acquisition perspective (just use ERA5 reanalysis).

As you can see in `ERA5StationOOTGDataGenerator`, this data generator has `xc` and `yc` which are of shape `[B, Nc, 2]` (because we only the spatial dimensions lat-lon) and `[B, Nc, dim_y]`. These represent the off-the-grid data and can be at arbitrary locations.  For the gridded variables, we have `xc_grid`, which is of shape `[B, lat, lon, 2]`, and `yc_grid`, which is of shape `[B, lat, lon, dim_y]`.

Note that in this example we are only interested in predicting the off-the-grid variable (t2m), and the off-the-grid variable is different to the on-the-grid one (t2m versus skin temperature). Depending on your exact use case, you might have to modify how you construct your batches such that the target incorporates a mix of on-the-grid and off-the-grid data.


3. > * how does the Gridded-TNP framework handle gridded data of different resolutions?

This is not explicitly implemented in the code, but it would be a fairly straightforward adaptation of the multi modal task. If your data is coming from the same modality (but different resolutions), you would need to make sure you are using the same `xy_encoder` and `y_encoder` (which can also be the identity). Then, similarly to how we are performing the multi modal task, you could have a separate grid for each data source (in this case each original resolution) that projects from the original resolution (which varies) to the fixed pseudo-token grid resolution.

> The shape of the values `yc[key]` is `[B, Nc, 1]` because that is the dimension of a single of the - in this data example - 6 different data variables, but they are just separated under different keys, and there is no notion of lat, lon, time, correct?

This is indeed correct. However, you do have the associated locations in `xc[key]`.

If you have more questions, we could also take this offline and discuss more your specific setup. Let me know if that would help.

**Comment by nilsleh** [comment]
@cddcam Thanks for the reply regarding the mix of station data and gridded data, I have come up with another test case file, to make it easier to step through the code:

```python
import torch
import tempfile
import os
import hiyapyco
from hydra.utils import instantiate
import xarray as xr
import numpy as np
import shutil

from tnp.utils.experiment_utils import deep_convert_dict, extract_config


def create_mock_era5_station_data(
    data_dir: str,
    gridded_data_vars: list,
    non_gridded_data_vars: list,
    time_steps: int = 24,
    lat_size: int = 181,
    lon_size: int = 360,
    num_stations: int = 50,
):
    """Create simple mock ERA5 data files and station coordinates for OOTG setup."""
    all_data_vars = list(set(gridded_data_vars + non_gridded_data_vars))

    os.makedirs(data_dir, exist_ok=True)
    
    # 1. Create simple ERA5 gridded data files
    # dates need to match time extend of train, val
    base_time = np.datetime64("2017-12-15T00:00:00")
    times = base_time + np.arange(time_steps) * np.timedelta64(1, "D")
    
    # Simple grid
    lats = np.linspace(90.0, -90.0, lat_size)
    lons = np.linspace(-180.0, 179.0, lon_size)
    
    # Create simple random data for each variable
    # station data gets subsampled in the generator
    data_arrays = {}
    for var in all_data_vars:
        data = np.random.randn(time_steps, lat_size, lon_size)
        data_arrays[var] = xr.DataArray(
            data,
            dims=["valid_time", "latitude", "longitude"],
            coords={"valid_time": times, "latitude": lats, "longitude": lons},
        )
    
    # Save dataset
    dataset = xr.Dataset(data_arrays)
    fname = os.path.join(data_dir, "era5_ootg_test_data.nc")
    dataset.to_netcdf(fname, engine="h5netcdf")
    
    # 2. Create station coordinates as PyTorch tensor (like hadisd_station_gridded.pt)
    # Generate random station coordinates: shape (num_stations, 2) for [lat, lon]
    np.random.seed(42)  # Reproducible
    station_coords = np.random.uniform(
        low=[-90, -180], 
        high=[90, 180], 
        size=(num_stations, 2)
    )
    
    # Convert to PyTorch tensor and save as .pt file
    station_coords_tensor = torch.tensor(station_coords, dtype=torch.float32)
    station_coords_fname = os.path.join(data_dir, "station_coords.pt")
    torch.save(station_coords_tensor, station_coords_fname)
    
    print(f"✓ Created ERA5 data: {fname}")
    print(f"✓ Created station coords: {station_coords_fname}")
    print(f"✓ Station coords shape: {station_coords_tensor.shape}")
    print(f"✓ Data vars: {all_data_vars}")
    
    return {
        "data_files": ["era5_ootg_test_data.nc"],
        "station_coords_fname": os.path.join(data_dir,"station_coords.pt"),
    }


def test_multimodal_tnp_encoder_from_configs():
    """Test MultiModalGriddedTNPEncoder using config files."""

    # Load and merge config files
    config_files = [
        "experiments/configs/models/ootg-swintnp.yml",
        "experiments/configs/models/grid_encoders/ootg-pt-ge.yml", 
        "experiments/configs/models/grid_decoders/mhca-gd.yml",
        "experiments/configs/generators/ootg-era5.yml",
        "experiments/configs/data/era5-nc-files.yml"
    ]

    raw_config = deep_convert_dict(
        hiyapyco.load(
            config_files, method=hiyapyco.METHOD_MERGE, usedefaultyamlloader=True
        )
    )

    # Create mock data
    temp_data_dir = tempfile.mkdtemp()
    gridded_data_vars = ["skt"]
    non_gridded_data_vars = ["t2m"]

    mock_data_info = create_mock_era5_station_data(
        temp_data_dir, 
        gridded_data_vars,
        non_gridded_data_vars,
        time_steps=24,
        lat_size=181,
        lon_size=360,
        num_stations=20
    )
    fnames = mock_data_info["data_files"]
    station_coords_fname = mock_data_info["station_coords_fname"]

    # Configure for testing
    raw_config.setdefault("data", {}).update(
        {
            "data_dir": temp_data_dir,
            "station_coords_fname": station_coords_fname,
            "train_fnames": fnames,
            "val_fnames": fnames,
        }
    )

    raw_config.setdefault("misc", {}).update(
        {
            "seed": 42,
            "num_workers": 1,
            "num_val_workers": 1,
            "logging": False,
            "pl_logging": False,
        }
    )

    raw_config.setdefault("params", {}).update(
        {
            "data_vars": gridded_data_vars + non_gridded_data_vars,
            "time_range": ["2018-01-01", "2018-01-01"],
            "val_time_range": ["2018-01-01", "2018-01-01"],
        }
    )

    # Extract and instantiate experiment
    experiment_config, _ = extract_config(raw_config, [])
    experiment = instantiate(experiment_config)

    print(f" Model instantiated: {type(experiment.model)}")
    print(f" Parameters: {sum(p.numel() for p in experiment.model.parameters()):,}")

    # Generate test batch and run forward pass
    test_batch = experiment.generators.train.generate_batch()

    with torch.no_grad():
        output = experiment.model(
            test_batch.xc, test_batch.yc, test_batch.xc_grid, test_batch.yc_grid, test_batch.xt
        )

    print(f"✓ Forward pass: {output.mean.shape}")
    shutil.rmtree(temp_data_dir, ignore_errors=True)

    return {"model": experiment.model, "output": output, "test_batch": test_batch}


if __name__ == "__main__":
    results = test_multimodal_tnp_encoder_from_configs()
```

Ultimately, I think the example I am most interested in is the following dataset setup, that I guess is a mix of MultiModal and Station Data so:
- several gridded variables that can have different resolutions 
- plus station data


Would the correct approach be that that the `grid_encoder` in the current ootg experiment setting [here](https://github.com/cambridge-mlg/gridded-tnp/blob/5b51367f412637ebf89da622270e8f75a280112c/experiments/configs/models/ootg-swintnp.yml#L10) would have to be wrapped with the [MultiModalGridEncoder](https://github.com/cambridge-mlg/gridded-tnp/blob/5b51367f412637ebf89da622270e8f75a280112c/experiments/configs/models/mm-swintnp.yml#L38) like it is used in the mm_config setup? But the per variable grid encoders would have to be `OOTGPseudoTokenGridEncoder` instead of the `PseudoTokenGridEncoderThroughTime`? 

I guess there would also have to be a new `GriddedTNP` class that combines the logic of the forward passes of [MultiModelGriddedTNPEncoder](https://github.com/cambridge-mlg/gridded-tnp/blob/5b51367f412637ebf89da622270e8f75a280112c/tnp/models/gridded_tnp.py#L110) and the [OOTGGriddedTNPEncoder](https://github.com/cambridge-mlg/gridded-tnp/blob/5b51367f412637ebf89da622270e8f75a280112c/tnp/models/gridded_tnp.py#L194) to handle that data case, right?
